geograypher/utils/numeric.py

```diff
@@ -30,40 +29,129 @@ def create_ramped_weighting(
 
 
 def compute_approximate_ray_intersection(
-    A: np.array, a: np.array, B: np.array, b: np.array, vis=False
+    a0: np.array,
+    a1: np.array,
+    b0: np.array,
+    b1: np.array,
+    clamp: bool = False,
+    plotter=None,
 ):
-    # https://palitri.com/vault/stuff/maths/Rays%18closest%20point.pdf
-    c = B - A
-
-    aa = np.dot(a, a)
-    ab = np.dot(a, b)
-    ac = np.dot(a, c)
-    bb = np.dot(b, b)
-    bc = np.dot(b, c)
-
-    denominator = aa * bb - ab * ab
-
-    a_scaler = (-ab * bc + ac * bb) / denominator
-    b_scaler = (ab * ac - bc * aa) / denominator
+    """
+    Given two line segments defined by 3D numpy.array pairs (a0, a1, b0, b1), return the
+    closest points on each segment and their distance. If clamp is True, then respect
+    the line segment ends. If clamp is False, then use the infinite rays.
 
-    D = A + a_scaler * a
-    E = B + b_scaler * b
+    Based on https://stackoverflow.com/questions/2824478/shortest-distance-between-two-line-segments
 
-    dist = np.linalg.norm(D - E)
-    # Check that it's in front of the camera
-    # TODO I'm not sure if this will handle co-linear cases, probably should check for that
-    valid = a_scaler > -1 and b_scaler > 0
+    Args:
+        a0 (np.ndarray): Start point of the first segment (shape: (3,)).
+        a1 (np.ndarray): End point of the first segment (shape: (3,)).
+        b0 (np.ndarray): Start point of the second segment (shape: (3,)).
+        b1 (np.ndarray): End point of the second segment (shape: (3,)).
+        clamp (bool, optional): If True, the closest points are clamped to the segment
+            endpoints. If False, the closest points may be anywhere along the infinite
+            lines.
+        plotter (pyvista.Plotter, optional): If provided, visualizes the segments
+            and closest points using pyvista.
 
-    if vis:
-        points = np.vstack([A, D, B, E, D, E])
+    Returns:
+        pA (np.ndarray): Closest point on the first segment or line (shape: (3,)),
+            or None if segments are parallel and overlap.
+        pB (np.ndarray): Closest point on the second segment or line (shape: (3,)),
+            or None if segments are parallel and overlap.
+        dist (float): The minimum distance between the two segments or lines.
+    """
 
-        plotter = pv.Plotter()
+    # Calculate vectors, normalized vectors, and denominator
+    A = a1 - a0
+    B = b1 - b0
+    magA = np.linalg.norm(A)
+    magB = np.linalg.norm(B)
+    # Normalized vectors
+    _A = A / magA
+    _B = B / magB
+    cross = np.cross(_A, _B)
+    denom = np.linalg.norm(cross) ** 2
+
+    # If lines are parallel (denom=0) test if lines overlap. If they don't
+    # overlap then there is a closest point solution. If they do overlap,
+    # there are infinite closest positions, but there is a closest distance
+    if not denom:
```

Comment:
Providing an intuition for this quantity could be helpful. Is the projection of connecting the two start points onto the first vector?

Reply:
No need to dig deeper if there's not a clear meaning to it. I doubt this code will be touched frequently and the tests are comprehensive.